### ISSUE TITLE
Bulk Update 10000+ records

### ISSUE THREAD

**Issue by dallisking** [simple-salesforce/simple-salesforce#383]
I'm having an issue with attempting to bulk update 10000+ records. Below is the code I'm running:

```
data = []
record_counter = 0
print("Total Records: {0}".format(len(results)))
for(col1, col2) in results:
    if record_counter < 10000:
        record = {}
        record['Id'] = col1
        record['score__c'] = col2
        data.append(record)
        record_counter += 1
    else:
        print("Updating {0} records".format(record_counter))
        sf.bulk.Health_Score__c.update(data)
        data = []
        record_counter = 0
```

The output from this is 
`
Total Records: 10457
Updating 10000 records
`
I would expect that after uploading the first batch of 10000, it would loop back to the top and continue on for the other 457 records, however the program simply ends after completing the first upload of 10000 (and no longer prints the results I've noticed...) I'm unfortunately not really much of a programmer. Is there a better way to be achieving this? Thanks so much for input!

**Comment by wyattshapiro** [comment]
@dallisking 
As you iterate through the for loop for the first 10k records, you will keep appending the 'record' dictionary to the 'data' list. 

Once you reach record 10,000, your code will not append that record, but instead will execute the else loop and run sf.bulk update for the first 10k records. Then, it resets the data variable to an empty list and the counter back down to 0.
**Issue 1:** You will always skip updating one record (namely record 10,000).

Then, you start from record 10001 iterating through the results until you hit record 10457. At that point, your 'for' loop will end.
**Issue 2:** The else statement that does the bulk update is never executed for the second batch because the for loop quits before the else statement can be executed.

In this case, I would recommend against using the bulk.update method while you are constructing the bulk data list since it might confuse you on what is happening. Instead, you could create a for loop that iterates through each record and adds the record to a list. Then, you could split that list into multiple lists each with a max length (ex. one list has 10k records, other list has 457 records). Finally, you could create a for loop that updates the records for each list you created.

This may not be the most efficient way to do this, but it separates out the logic of batching up your records from uploading them!

**Comment by jon-wobken** [comment]
@dallisking - Resolved in v1.10 with multi-threading. New batch_size argument is supported on delete/insert/upsert/update/hard_delete with default of 10,000
